typing/typedecl: split and factorize fixpoint computations for variance and immediacy

gasche opened the pull request:
Many properties of (potentially mutually-recursive) type declarations need to be computed by an iterative fixpoint. The OCaml type-checker currently knows of two, variance of type parameters, and immediacy (the `immediate` attribute, see [documentation](http://caml.inria.fr/pub/docs/manual-ocaml-4.07/extn.html#sec260)).

Currently variance and immediacy are computed in a single fixpoint loop, iterating on both as long as one of them has not reached a stable result. This is a bit wasteful (one of them is going to be recomputed more than necessary), and it also makes the code non-modular and annoying to extend if we want to add new properties (precisely my goal, with a new fixpoint computation for `unboxed` type declarations).

This PR first splits the single loop into two independent loops, and then it implements a new generic loop (on an abstract notion of "typedecl property" with suitable operations) and ports both variance and immediacy to use this new, more generic framework.

gasche committed:
typedecl: split the fixpoint computations for variance and immediacy

This change makes the code simpler and easier to extend. It also
avoids useless computation: the two fixpoints may take a different
number of iterations to converge, and the previous code would compute
both properties as long as one of them hadn't converged.

gasche committed:
introduce a generic structure for "type-declaration properties"

Also included are a generic combinator, and a port of the variance
computation to the generic scheme -- without both those things, we
wouldn't know we have the right (internal) interface.

gasche committed:
typedecl: port immediacy computation to the new generic scheme

gasche commented:
cc @garrigue, who wrote the original code, #188 the PR that added immediacy, and @Armael, who might be interested.

gasche committed:
Changes entry

Armael commented on typing/typedecl.ml:
```diff
@@ -1135,55 +1135,96 @@ let compute_immediacy env tdecl =
+  let rec compute_fixpoint props =
+    let new_decls =
+      List.map2 (fun (id, decl) prop ->
+          (id, property.update_decl decl prop ))
```
Typo: Extra space here

gasche replied:
Fixed, thanks.

Armael commented:
I haven't tried to understand in all details what the code is doing, but from a high-level perspective this looks like a nice refactoring.

gasche committed:
add Types.Variance.eq

gasche commented:
@Armael no pressure, but I can't tell from your message whether you plan to do an actual review, you plan to not do an actual review, or you don't know yet. (Typical observable outcomes are one of: (1) a merge decision, (2) a request for changes before a merge decision can be made, or (3) an indication that a review will not be conducted (typically by lack of time or expertise) and often a passing of bucket to someone else.)

Armael commented:
Right, sorry. I'll try to do an actual review in the next few days, unless someone else beats me to it.

Armael commented on typing/typedecl.ml:
```diff
@@ -1135,55 +1135,96 @@ let compute_immediacy env tdecl =
+  check : Env.t -> Ident.t -> decl -> 'req -> unit;
+}
+
+let compute_property property env decls required =
```
If I understand correctly, `decls` and `required` must have the same length (and represent a mapping decl -> required)? It would be good to make that explicit in a comment.

Armael replied:
A nice property of your refactoring is that the type of `compute_property` is now of the form `... -> decls -> decls` (where `decls = (Ident.t * decl) list`, i.e. it takes a list of declarations and updates it with extra information.
This means that it composes nicely: we can chain calls to `compute_property` with different properties, as you do later on.

This is intense nitpicking, but I think it would be nice to make that very explicit by taking the `decls` argument last, introduce a `type decls = (Ident.t * decl) list` alias, and add type annotations to the arguments of `compute_property`.

gasche replied:
I did add an annotation and a documentation comment for `'req` and `required`. On the other hand, I tried your suggestion to have `decls` be the last parameter, and I don't like the result: passing `required` before `decls` doesn't make much sense (try formulating a sequence that describe what the function does, for example, you will see that you really want to respect the Pi-types restriction that dependent objects come after what they depend on).

We could still have piping in the big `transl_type_decl` function if I declared per-property wrappers for the fixpoint, that would also take care of computing the requirements. I'll try this.

Armael replied:
I like the idea.

gasche replied:
I just pushed a new commit that uses pipes in the top-dog function.

Armael replied:
LGTM

Armael commented on typing/typedecl.ml:
```diff
@@ -1135,55 +1135,96 @@ let compute_immediacy env tdecl =
-      new_decls required
+  compute_fixpoint props
+
+type variance_req = (bool * bool * bool) list * Location.t
```
Do you know what these booleans (and location) mean? If yes, could that be explained in a comment?

gasche replied:
I don't really care what they mean (I think they come from types.mli:Variance), but your question made me realize that we can probably get rid of the `Location.t` parameter: the current code gets it from somewhere else, but the same location can be found in `decl`.

gasche replied:
I could get rid of the locations in this part of the code, but not in the family of `compute_variance_foo` functions. I just pushed a new commit.

Armael reviewed (approved):
I have checked that the new code is equivalent to the old code. I also agree that the refactoring is nicer and more flexible. Hence, I approve the PR; it should be safe to merge it as it is now.

Additionally, I have added a few comments on the code that are mostly nitpicking and you are free to address them or not. 

I have a last question: could you explain a bit what is the informal semantics of the `required` list? With only two instantiations (at the moment), and only one that uses it, I'm having a hard time understanding whether it follows a more general principle or is specific to the variance computation.
Do you have more insight from your work on unboxed annotations?

gasche commented:
Thanks for the review! 

Regarding `'req` and `required`: some properties come with an expression of intent from the user: it's the difference between `type 'a t = ..` (infer the best variance you can) and `type +'a t = ...` (infer the best variance, but it should be at least `+`). `required` represents this requirement, for each declaration in the declaration list. The previous implement would pass requirements for variance, and check that they are respected by the final solution at the end of the fixpoint loop; this is what the `required` parameter (of type `'req`) and the `check : env -> id -> 'prop -> 'req -> unit` operation on properties are doing. 

Properties for which the implementation doesn't currently have such a notion can use `unit` as their requirement type. This is the case for immediacies in this PR, which follows the current implementation. In fact, there is a notion of requirement for immediacies (if the attribute is used) that is checked someplace else in the code, so I could consider moving it that here -- at type `bool option`, I suppose.

I haven't thought yet about how to move our `[@unboxed]` check on top of this infrastructure (@rlepigre and myself implemented our own fixpoint, and I have yet to rebase on top of this later interface), but I don't think we would use non-`unit` requirements; there is a global requirement coming from the OCaml runtime, that each type (introduced by a type declaration) must be separable, not a per-declaration requirement that would come from the user.

gasche committed:
rename Typedecl.compute_variance_decls into compute_variance_class_decls

gasche committed:
typedecl: remove the location from variance requirements

The location can always be reconstructed from `decl.type_loc`. Note
however that variance-computation functions, in particular
`compute_variance_extension` and its callee (`compute_variance_gadt`,
etc.) need to be passed both the requirement and the location, despite
the fact that they also take a `decl` parameter: there are other
use-cases where the location is not the location of the type
declaration -- when processing an extension to an existing type
declaration.

gasche committed:
typedecl: refactor variance requirements to present a unified pipable interface

Armael commented:
LGTM. Can merge when CI passes.

yallop commented:
> it's the difference between `type 'a t = ..` (infer the best variance you can) and `type +'a t = ...` (infer the best variance, but it should be at least `+`).

I think that's how it should work.  But at the moment adding `+` annotations to GADT-style definitions can actually change the variance from invariant to covariant.

garrigue commented:
Indeed, inference mode is limited to the most simple case: variable type parameter in a public type definition, that does not appear in a constraint, directly or via one of the GADT cases. Otherwise the annotation is required to get any form of variance. The reason is that we need to know the variance of constrained parameters in other to infer that of the variables that appear in them (i.e. we only infer the variance for type variables, not type expressions).
